packages/convex-helpers/server/zod4.ts

```diff
@@ -0,0 +1,642 @@
+    }
+    return v.object(out);
+  }
+
+  if (zodValidator instanceof z.ZodUnion) {
+    const options = (zodValidator as any).options as z.ZodTypeAny[];
+    if (Array.isArray(options) && options.length > 0) {
+      const members = options.map((opt) => zodToConvexInternal(opt, visited));
+      return (makeUnion(members as [GenericValidator, GenericValidator, ...GenericValidator[]]) as unknown) as GenericValidator;
+    }
+    return v.any();
+  }
+  if (zodValidator instanceof z.ZodDiscriminatedUnion) {
+    const options: Iterable<z.ZodTypeAny> =
+      ((zodValidator as any).def?.options as z.ZodTypeAny[]) ||
+      ((zodValidator as any).def?.optionsMap?.values?.() as Iterable<z.ZodTypeAny> | undefined) ||
+      [];
+    const arr = Array.isArray(options) ? options : Array.from(options);
+    if (arr.length >= 1) {
+      const vals = arr.map((opt) => zodToConvexInternal(opt, visited));
+      return (makeUnion(vals as [GenericValidator, ...GenericValidator[]]) as unknown) as GenericValidator;
+    }
+    return v.any();
+  }
+  if (zodValidator instanceof z.ZodTuple) {
+    const items = ((zodValidator as any).def?.items as z.ZodTypeAny[] | undefined) ?? [];
+    if (items.length > 0) {
+      const members = items.map((it) => zodToConvexInternal(it, visited));
+      const unionized = makeUnion(members);
+      return v.array(unionized as unknown as any);
+    }
+    return v.array(v.any());
+  }
+  if (zodValidator instanceof z.ZodLazy) {
+    try {
+      const getter = (zodValidator as any).def?.getter as (() => unknown) | undefined;
+      if (getter) return zodToConvexInternal(getter() as unknown as z.ZodTypeAny, visited);
+    } catch {
+      // ignore resolution errors
+    }
+    return v.any();
+  }
+  if (zodValidator instanceof z.ZodLiteral) {
+    const val = (zodValidator as any).value as string | number | boolean | bigint | null;
+    return v.literal(val as any);
+  }
+  if (zodValidator instanceof z.ZodEnum) {
+    const options = (zodValidator as any).options as unknown[];
+    const literals = options.map((o) => v.literal(o as any)) as unknown as GenericValidator[];
+    if (literals.length === 0) return v.any();
+    return makeUnion(literals) as unknown as GenericValidator;
+  }
+  if (zodValidator instanceof z.ZodRecord) {
+    const valueType = (zodValidator as any).valueType as z.ZodTypeAny | undefined;
+    const vVal = valueType ? zodToConvexInternal(valueType, visited) : v.any();
+    return v.record(v.string(), asValidator(vVal));
+  }
+  if (zodValidator instanceof z.ZodReadonly) {
+    return zodToConvexInternal((zodValidator as any).innerType as z.ZodTypeAny, visited);
+  }
+  if (zodValidator instanceof z.ZodTransform) {
+    const inner = (zodValidator as any).def?.schema as z.ZodTypeAny | undefined;
+    return inner ? zodToConvexInternal(inner, visited) : v.any();
+  }
+  return v.any();
+}
+
+export function zodToConvex<Z extends z.ZodTypeAny | ZodValidator>(zodSchema: Z) {
```

Comment:
Why is this function also accepting records? The homologous function from our Zed 3 support only accepts Zod 3 types, and `zodToConvexFields` is used to support objects of Zod types.